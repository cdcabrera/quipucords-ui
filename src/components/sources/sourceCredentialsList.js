import React from 'react';
import PropTypes from 'prop-types';
import { Grid, Icon } from 'patternfly-react';
import { apiTypes } from '../../constants/apiConstants';

const SourceCredentialsList = ({ source }) => {
  const credentials =
    (source[apiTypes.API_RESPONSE_SOURCE_CREDENTIALS] && [...source[apiTypes.API_RESPONSE_SOURCE_CREDENTIALS]]) || [];

  credentials.sort((item1, item2) =>
    item1[apiTypes.API_RESPONSE_SOURCE_CREDENTIALS_NAME].localeCompare(
      item2[apiTypes.API_RESPONSE_SOURCE_CREDENTIALS_NAME]
    )
  );

  return (
    <Grid fluid>
      {credentials.map(credential => (
        <Grid.Row key={credential[apiTypes.API_RESPONSE_SOURCE_CREDENTIALS_NAME]}>
          <Grid.Col xs={12} sm={4}>
            <span>
              <Icon type="fa" name="id-card" />
              &nbsp; {credential[apiTypes.API_RESPONSE_SOURCE_CREDENTIALS_NAME]}
            </span>
          </Grid.Col>
        </Grid.Row>
      ))}
    </Grid>
  );
};

SourceCredentialsList.propTypes = {
  source: PropTypes.shape({
    credentials: PropTypes.array
  }).isRequired
};

export { SourceCredentialsList as default, SourceCredentialsList };
