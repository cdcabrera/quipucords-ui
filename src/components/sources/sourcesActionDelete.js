import { reduxTypes, store } from '../../redux'
import { helpers } from '../../common'
import React from 'react'

const onDelete = source => {
  const { deleteSource } = this.props;

  const onConfirm = () => {
    store.dispatch({
      type: reduxTypes.confirmationModal.CONFIRMATION_MODAL_HIDE
    });

    deleteSource(source[apiTypes.API_RESPONSE_SOURCE_ID]).then(
      () => {
        store.dispatch({
          type: reduxTypes.toastNotifications.TOAST_ADD,
          alertType: 'success',
          message: (
            <span>
                Deleted source <strong>{source[apiTypes.API_RESPONSE_SOURCE_NAME]}</strong>.
              </span>
          )
        });

        store.dispatch({
          type: reduxTypes.view.DESELECT_ITEM,
          viewType: reduxTypes.view.SOURCES_VIEW,
          item: source
        });

        store.dispatch({
          type: reduxTypes.sources.UPDATE_SOURCES
        });
      },
      error => {
        store.dispatch({
          type: reduxTypes.toastNotifications.TOAST_ADD,
          alertType: 'danger',
          header: 'Error',
          message: helpers.getMessageFromResults(error).message
        });
      }
    );
  };

  store.dispatch({
    type: reduxTypes.confirmationModal.CONFIRMATION_MODAL_SHOW,
    title: 'Delete Source',
    heading: (
      <span>
          Are you sure you want to delete the source <strong>{source[apiTypes.API_RESPONSE_SOURCE_NAME]}</strong>?
        </span>
    ),
    confirmButtonText: 'Delete',
    onConfirm
  });
};

const onEdit = source => {
  dispatch({
    type: reduxTypes.sources.EDIT_SOURCE_SHOW,
    source
  });
};

const onScan = source => {
  dispatch({
    type: reduxTypes.scans.EDIT_SCAN_SHOW,
    sources: [source]
  });
};
