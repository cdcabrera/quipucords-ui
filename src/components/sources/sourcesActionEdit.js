import { reduxTypes, store } from '../../redux'
import { helpers } from '../../common'
import React from 'react';


const onEdit = source => {
  dispatch({
    type: reduxTypes.sources.EDIT_SOURCE_SHOW,
    source
  });
};

const onScan = source => {
  dispatch({
    type: reduxTypes.scans.EDIT_SCAN_SHOW,
    sources: [source]
  });
};
