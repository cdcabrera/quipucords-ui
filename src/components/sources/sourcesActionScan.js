import React from 'react';
import { reduxTypes, storeHooks } from '../../redux';
import { helpers } from '../../common';

const onUseScan = ({ useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch } = {}) => {
  const dispatch = useAliasDispatch();

  return source => {
    dispatch({
      type: reduxTypes.scans.EDIT_SCAN_SHOW,
      sources: [source]
    });
  };
};

