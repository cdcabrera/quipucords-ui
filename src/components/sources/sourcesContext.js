import React, { useState } from 'react';
import { useShallowCompareEffect, useUnmount } from 'react-use';
import { reduxActions, reduxTypes, storeHooks } from '../../redux';
import { apiTypes } from '../../constants/apiConstants';
import { helpers } from '../../common';

/**
 * On Delete confirmation, and action.
 *
 * @param {object} options
 * @param {Function} options.deleteSource
 * @param {Function} options.useDispatch
 * @param {Function} options.useSelector
 * @param {Function} options.useSelectorsResponse
 * @returns {Function}
 */
const useOnDelete = ({
  deleteSource = reduxActions.sources.deleteSource,
  useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch,
  useSelector: useAliasSelector = storeHooks.reactRedux.useSelector,
  useSelectorsResponse: useAliasSelectorsResponse = storeHooks.reactRedux.useSelectorsResponse
} = {}) => {
  const dispatch = useAliasDispatch();
  const sourceToDelete = useAliasSelector(({ sources }) => sources?.confirmDelete?.source, {});
  const { error, fulfilled, message } = useAliasSelectorsResponse(({ sources }) => sources?.deleted);
  const { [apiTypes.API_RESPONSE_SOURCE_ID]: sourceId, [apiTypes.API_RESPONSE_SOURCE_NAME]: sourceName } =
    sourceToDelete;

  useShallowCompareEffect(() => {
    if (sourceId) {
      dispatch([
        {
          type: reduxTypes.confirmationModal.CONFIRMATION_MODAL_HIDE
        }
      ]);

      deleteSource(sourceId)(dispatch);
    }
  }, [sourceId, deleteSource, dispatch]);

  useShallowCompareEffect(() => {
    if (fulfilled) {
      dispatch([
        {
          type: reduxTypes.toastNotifications.TOAST_ADD,
          alertType: 'success',
          message: (
            <React.Fragment>
              Deleted source <strong>{sourceName}</strong>.
            </React.Fragment>
          )
        },
        {
          type: reduxTypes.view.DESELECT_ITEM,
          viewType: reduxTypes.view.SOURCES_VIEW,
          item: sourceToDelete
        },
        {
          type: reduxTypes.sources.UPDATE_SOURCES
        }
      ]);
    }

    if (error) {
      dispatch({
        type: reduxTypes.toastNotifications.TOAST_ADD,
        alertType: 'danger',
        header: 'Error',
        message
      });
    }
  }, [error, fulfilled, message, dispatch]);

  return source => {
    dispatch({
      type: reduxTypes.confirmationModal.CONFIRMATION_MODAL_SHOW,
      title: 'Delete Source',
      heading: (
        <span>
          Are you sure you want to delete the source <strong>{source[apiTypes.API_RESPONSE_SOURCE_NAME]}</strong>?
        </span>
      ),
      confirmButtonText: 'Delete',
      onConfirm: () =>
        dispatch({
          type: reduxTypes.sources.CONFIRM_DELETE_SOURCE,
          source
        })
    });
  };
};

/**
 * On edit a source
 *
 * @param {object} options
 * @param {Function} options.useDispatch
 * @returns {Function}
 */
const useOnEdit = ({ useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch } = {}) => {
  const dispatch = useAliasDispatch();

  return source => {
    dispatch({
      type: reduxTypes.sources.EDIT_SOURCE_SHOW,
      source
    });
  };
};

/**
 * On scan a source
 *
 * @param {object} options
 * @param {Function} options.useDispatch
 * @returns {Function}
 */
const useOnScan = ({ useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch } = {}) => {
  const dispatch = useAliasDispatch();

  return source => {
    dispatch({
      type: reduxTypes.scans.EDIT_SCAN_SHOW,
      sources: [source]
    });
  };
};

/**
 * Poll sources data for pending results.
 *
 * @param {object} options
 * @param {number} options.pollInterval
 * @param {Function} options.useSelector
 * @returns {Function}
 */
const usePoll = ({
  pollInterval = helpers.POLL_INTERVAL,
  useSelector: useAliasSelector = storeHooks.reactRedux.useSelector
} = {}) => {
  const [timer, setTimer] = useState();
  const [updatePoll, setUpdatePoll] = useState(0);
  const updatedSources = useAliasSelector(({ sources }) => sources?.view?.data?.results, []);
  const shouldUpdate = updatedSources.find(
    ({ connection }) =>
      connection.status === 'created' || connection.status === 'pending' || connection.status === 'running'
  );

  useUnmount(() => {
    window.clearTimeout(timer);
  });

  useShallowCompareEffect(() => {
    if (shouldUpdate) {
      setTimer(
        window.setTimeout(() => {
          setUpdatePoll(helpers.getCurrentDate().getTime());
        }, pollInterval)
      );
    }
  }, [pollInterval, shouldUpdate]);

  return updatePoll;
};

const context = {
  useOnDelete,
  useOnEdit,
  useOnScan,
  usePoll
};

export { context as default, context, useOnDelete, useOnEdit, useOnScan, usePoll };
