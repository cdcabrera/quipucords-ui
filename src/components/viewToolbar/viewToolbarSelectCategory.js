import React from 'react';
import PropTypes from 'prop-types';
// import { useShallowCompareEffect } from 'react-use';
import { FilterIcon } from '@patternfly/react-icons';
import { reduxTypes, storeHooks } from '../../redux';
import { DropdownSelect } from '../dropdownSelect/dropdownSelect';
import { translate } from '../i18n/i18n';
import { useView } from '../view/viewContext';
// import filterFields from '../credentials/credentialsToolbar'

/**
 * On select update category.
 *
 * @param {string} viewId
 * @param {object} options
 * @param {Function} options.useDispatch
 * @returns {Function}
 */
const useOnSelect = (viewId, { useDispatch: useAliasDispatch = storeHooks.reactRedux.useDispatch } = {}) => {
  const dispatch = useAliasDispatch();

  return ({ value = null } = {}) => {
    dispatch([
      {
        type: reduxTypes.view.SET_FILTER,
        viewId,
        currentFilterCategory: value
      }
    ]);
  };
};

/**
 * Return filtered category options, current, and initial value.
 *
 * @param {object} options
 * @param {Array} options.categoryOptions
 * @param {Function} options.useProduct
 * @param {Function} options.useProductToolbarConfig
 * @param {Function} options.useSelector
 * @returns {object}
 */
/*
const useSelectCategoryOptions = ({
  // categoryOptions = toolbarFieldOptions,
  // useProduct: useAliasProduct = useProduct,
  useView: useAliasView = useView,
  // useProductToolbarConfig: useAliasProductToolbarConfig = useProductToolbarConfig,
  useSelector: useAliasSelector = storeHooks.reactRedux.useSelector
} = {}) => {
  const { viewId, config } = useAliasView();
  const { toolbar: options } = config;

  const { currentCategory } = useAliasSelector(({ view }) => view.filters?.[viewId], {});

  /*
  let initialCategory;

  const updatedOptions = options.map(option => {
    const { selected, value } = option;

    if (currentCategory === undefined && selected) {
      initialCategory = value;
    }

    return {
      ...option,
      selected: (currentCategory === undefined && selected) || currentCategory === value
    };
  });
  * /

  return {
    currentCategory,
    options
    // initialCategory,
    // options: updatedOptions
  };
};
 */

/**
 * Select available filter categories.
 *
 * @param {object} props
 * @param {Array} props.options
 * @param {Function} props.t
 * @param {Function} props.useOnSelect
 * @param {Function} props.useSelectCategoryOptions
 * @param {string} props.viewId
 * @returns {React.ReactNode}
 */
const ViewToolbarSelectCategory = ({
  // options,
  t,
  useOnSelect: useAliasOnSelect,
  // useSelector: useAliasSelector,
  // useSelectCategoryOptions: useAliasSelectCategoryOptions,
  useSelector: useAliasSelector,
  useView: useAliasView
  // viewId
  // t,
  // useOnSelect: useAliasOnSelect,
  // useSelectCategoryOptions: useAliasSelectCategoryOptions
}) => {
  // const selectedOption = useAliasSelector(({ viewOptions }) => viewOptions?.[viewId]?.currentCategory, undefined);
  const { viewId, config } = useAliasView();
  const options = config.toolbar.filterFields;
  const onSelect = useAliasOnSelect(viewId);

  const selectedOption = useAliasSelector(({ view }) => view?.filters?.[viewId]?.currentFilterCategory);
  // const { currentCategory } = useAliasSelector(({ view }) => view.filters?.[viewId], {});
  // const { categoryFields, initialCategory } = useAliasSelectCategoryOptions(viewId, options);

  /*
  useShallowCompareEffect(() => {
    if (initialCategory) {
      onSelect({ value: initialCategory });
    }
  }, [initialCategory, onSelect]);
  */

  return (
    <DropdownSelect
      ariaLabel={t('toolbar.label', { context: ['placeholder', 'filter'] })}
      placeholder={t('toolbar.label', { context: ['placeholder', 'filter'] })}
      options={options}
      onSelect={onSelect}
      selectedOptions={selectedOption}
      // selectedOptions={currentCategory}
      toggleIcon={<FilterIcon />}
    />
  );
};

/**
 * Prop types.
 *
 * @type {{useOnSelect: Function, t: Function, useSelectCategoryOptions: Function}}
 */
ViewToolbarSelectCategory.propTypes = {
  /*
  options: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
      value: PropTypes.any,
      selected: PropTypes.bool
    })
  ),
   */
  t: PropTypes.func,
  useOnSelect: PropTypes.func,
  useSelector: PropTypes.func,
  // useSelectCategoryOptions: PropTypes.func,
  useView: PropTypes.func
};

/**
 * Default props.
 *
 * @type {{useOnSelect: Function, t: Function, useSelectCategoryOptions: Function}}
 */
ViewToolbarSelectCategory.defaultProps = {
  // options: [],
  t: translate,
  useOnSelect,
  useSelector: storeHooks.reactRedux.useSelector,
  // useSelectCategoryOptions: Function.prototype,
  // useSelectCategoryOptions,
  useView
};

export { ViewToolbarSelectCategory as default, ViewToolbarSelectCategory, useOnSelect };
