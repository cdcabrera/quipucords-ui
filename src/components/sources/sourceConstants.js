import React from 'react';
import { translate } from '../i18n/i18n';
// import { dictionary } from '../../constants/dictionaryConstants';
import { ViewToolbarFieldFilter } from '../viewToolbar/viewToolbarFieldFilter';

const SourceFilterFields = [
  {
    title: () => translate('toolbar.label', { context: ['option', 'search_by_name'] }),
    value: 'search_by_name',
    component: function SearchName(props) {
      return <ViewToolbarFieldFilter queryType="search_by_name" {...props} />;
    },
    selected: true
  },
  {
    title: () => translate('toolbar.label', { context: ['option', 'search_credentials_by_name'] }),
    value: 'search_credentials_by_name',
    component: function SearchName(props) {
      return <ViewToolbarFieldFilter queryType="search_credentials_by_name" {...props} />;
    }
  }
  /*
  {
    id: 'search_by_name',
    title: 'Name',
    placeholder: 'Filter by Name',
    filterType: 'text'
  },
  {
    id: 'search_credentials_by_name',
    title: 'Credential',
    placeholder: 'Filter by Credential Name',
    filterType: 'text'
  },
  {
    id: 'source_type',
    title: 'Source Type',
    placeholder: 'Filter by Source Type',
    filterType: 'select',
    filterValues: [
      { title: dictionary.network, id: 'network' },
      { title: dictionary.satellite, id: 'satellite' },
      { title: dictionary.vcenter, id: 'vcenter' }
    ]
  }
  */
];

/**
 * ID: Enum with the following possible values [name, source_type, most_recent_connect_scan__start_time]
 */
const SourceSortFields = [
  {
    title: () => translate('toolbar.label', { context: ['option', 'name'] }),
    value: 'name',
    isNumeric: false,
    selected: true
  },
  {
    title: () => translate('toolbar.label', { context: ['option', 'source_type'] }),
    value: 'source_type',
    isNumeric: false
  },
  {
    title: () => translate('toolbar.label', { context: ['option', 'most_recent'] }),
    value: 'most_recent_connect_scan__start_time',
    isNumeric: true,
    sortAscending: false
  }
];

export { SourceFilterFields, SourceSortFields };
