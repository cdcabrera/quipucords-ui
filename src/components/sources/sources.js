import React from 'react';
import PropTypes from 'prop-types';
import { Alert, Button, EmptyState, ListView, Modal, Spinner } from 'patternfly-react';
import _isEqual from 'lodash/isEqual';
import _size from 'lodash/size';
import { connect, reduxActions, reduxTypes, store } from '../../redux';
import helpers from '../../common/helpers';
import ViewToolbar from '../viewToolbar/viewToolbar';
import ViewPaginationRow from '../viewPaginationRow/viewPaginationRow';
import SourcesEmptyState from './sourcesEmptyState';
import SourceListItem from './sourceListItem';
import { SourceFilterFields, SourceSortFields } from './sourceConstants';
import { apiTypes } from '../../constants/apiConstants';

class Sources extends React.Component {
  componentDidMount() {
    const { getSources, viewOptions } = this.props;

    getSources(helpers.createViewQueryObject(viewOptions));
  }

  componentDidUpdate(prevProps) {
    const { getSources, updateSources, viewOptions } = this.props;

    const prevQuery = helpers.createViewQueryObject(prevProps.viewOptions);
    const nextQuery = helpers.createViewQueryObject(viewOptions);

    if (updateSources || !_isEqual(prevQuery, nextQuery)) {
      getSources(nextQuery);
    }
  }

  onShowAddSourceWizard = () => {
    store.dispatch({
      type: reduxTypes.sources.CREATE_SOURCE_SHOW
    });
  };

  onScanSources = () => {
    const { viewOptions } = this.props;

    store.dispatch({
      type: reduxTypes.scans.EDIT_SCAN_SHOW,
      sources: viewOptions.selectedItems
    });
  };

  onRefresh = () => {
    store.dispatch({
      type: reduxTypes.sources.UPDATE_SOURCES
    });
  };

  onClearFilters = () => {
    store.dispatch({
      type: reduxTypes.viewToolbar.CLEAR_FILTERS,
      viewType: reduxTypes.view.SOURCES_VIEW
    });
  };

  renderSourceActions() {
    const { viewOptions } = this.props;

    return (
      <div className="form-group">
        <Button bsStyle="primary" onClick={this.onShowAddSourceWizard}>
          Add
        </Button>
        <Button
          disabled={!viewOptions.selectedItems || viewOptions.selectedItems.length === 0}
          onClick={this.onScanSources}
        >
          Scan
        </Button>
      </div>
    );
  }

  renderPendingMessage() {
    const { pending } = this.props;

    if (pending) {
      return (
        <Modal bsSize="lg" backdrop={false} show animation={false}>
          <Modal.Body>
            <Spinner loading size="lg" className="blank-slate-pf-icon" />
            <div className="text-center">Loading...</div>
          </Modal.Body>
        </Modal>
      );
    }

    return null;
  }

  renderSourcesList(sources) {
    const { lastRefresh } = this.props;

    if (sources.length) {
      return (
        <ListView className="quipicords-list-view">
          {sources.map(source => (
            <SourceListItem item={source} key={source[apiTypes.API_RESPONSE_SOURCE_ID]} lastRefresh={lastRefresh} />
          ))}
        </ListView>
      );
    }

    return (
      <EmptyState className="list-view-blank-slate">
        <EmptyState.Title>No Results Match the Filter Criteria</EmptyState.Title>
        <EmptyState.Info>The active filters are hiding all items.</EmptyState.Info>
        <EmptyState.Action>
          <Button bsStyle="link" onClick={this.onClearFilters}>
            Clear Filters
          </Button>
        </EmptyState.Action>
      </EmptyState>
    );
  }

  render() {
    const { error, errorMessage, lastRefresh, pending, sources, viewOptions } = this.props;

    if (error) {
      return (
        <EmptyState>
          <Alert type="error">
            <span>Error retrieving sources: {errorMessage}</span>
          </Alert>
          {this.renderPendingMessage()}
        </EmptyState>
      );
    }

    if (pending && !sources.length) {
      return <div className="quipucords-view-container">{this.renderPendingMessage()}</div>;
    }

    if (sources.length || _size(viewOptions.activeFilters)) {
      return (
        <div className="quipucords-view-container">
          <ViewToolbar
            viewType={reduxTypes.view.SOURCES_VIEW}
            filterFields={SourceFilterFields}
            sortFields={SourceSortFields}
            onRefresh={this.onRefresh}
            lastRefresh={lastRefresh}
            actions={this.renderSourceActions()}
            itemsType="Source"
            itemsTypePlural="Sources"
            selectedCount={viewOptions.selectedItems.length}
            {...viewOptions}
          />
          <ViewPaginationRow viewType={reduxTypes.view.SOURCES_VIEW} {...viewOptions} />
          <div className="quipucords-list-container">{this.renderSourcesList(sources)}</div>
          {this.renderPendingMessage()}
        </div>
      );
    }

    return <SourcesEmptyState onAddSource={this.onShowAddSourceWizard} />;
  }
}

Sources.propTypes = {
  error: PropTypes.bool,
  errorMessage: PropTypes.string,
  getSources: PropTypes.func,
  lastRefresh: PropTypes.number,
  pending: PropTypes.bool,
  sources: PropTypes.array,
  updateSources: PropTypes.bool,
  viewOptions: PropTypes.object
};

Sources.defaultProps = {
  error: false,
  errorMessage: null,
  getSources: helpers.noop,
  lastRefresh: 0,
  pending: false,
  sources: [],
  updateSources: false,
  viewOptions: {}
};

const mapDispatchToProps = dispatch => ({
  getSources: queryObj => dispatch(reduxActions.sources.getSources(queryObj))
});

const mapStateToProps = state =>
  Object.assign({}, state.sources.view, { viewOptions: state.viewOptions[reduxTypes.view.SOURCES_VIEW] });

const ConnectedSources = connect(
  mapStateToProps,
  mapDispatchToProps
)(Sources);

export { ConnectedSources as default, ConnectedSources, Sources };
