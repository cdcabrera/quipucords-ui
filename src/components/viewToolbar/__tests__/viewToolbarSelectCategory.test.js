import React from 'react';
import { ViewToolbarSelectCategory, useOnSelect } from '../viewToolbarSelectCategory';
import { API_QUERY_TYPES } from '../../../constants/apiConstants';
import { CONFIG as sourcesConfig } from '../../sources/sources';
import { store } from '../../../redux/store';

describe('ViewToolbarSelectCategory Component', () => {
  let mockDispatch;

  beforeEach(() => {
    mockDispatch = jest.spyOn(store, 'dispatch').mockImplementation((type, data) => ({ type, data }));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render a basic component', async () => {
    const props = {
      useView: () => ({ viewId: sourcesConfig.viewId, config: { toolbar: sourcesConfig.toolbar } })
    };
    const component = await shallowHookComponent(<ViewToolbarSelectCategory {...props} />);

    expect(component).toMatchSnapshot('basic');
  });

  it('should handle updating the view query through redux state with a component', async () => {
    const props = {
      useView: () => ({ viewId: sourcesConfig.viewId, config: { toolbar: sourcesConfig.toolbar } })
    };

    const component = await mountHookComponent(<ViewToolbarSelectCategory {...props} />);

    component.find('button').simulate('click');
    component.update();
    component.find('button.pf-c-select__menu-item').first().simulate('click');

    expect(mockDispatch.mock.calls).toMatchSnapshot('dispatch type, component');
  });

  it('should handle updating the view query through redux state with a hook', async () => {
    const options = {};
    const { result: onSelect } = await shallowHook(() => useOnSelect('lorem filter', options));

    onSelect({ value: 'dolor sit' });
    expect(mockDispatch.mock.calls).toMatchSnapshot('dispatch type, hook');
  });
});
