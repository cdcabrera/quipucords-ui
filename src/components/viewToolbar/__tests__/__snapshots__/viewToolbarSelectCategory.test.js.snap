// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewToolbarSelectCategory Component should handle updating the view query through redux state with a component: dispatch type, component 1`] = `
Array [
  Array [
    Array [
      Object {
        "currentFilterCategory": "search_by_name",
        "type": "SET_FILTER",
        "viewId": undefined,
      },
    ],
  ],
  Array [
    Array [
      Object {
        "currentFilterCategory": "search_by_name",
        "type": "SET_FILTER",
        "viewId": undefined,
      },
    ],
  ],
]
`;

exports[`ViewToolbarSelectCategory Component should handle updating the view query through redux state with a hook: dispatch type, hook 1`] = `
Array [
  Array [
    Array [
      Object {
        "currentFilterCategory": "dolor sit",
        "type": "SET_FILTER",
        "viewId": undefined,
      },
    ],
  ],
]
`;

exports[`ViewToolbarSelectCategory Component should render a basic component: basic 1`] = `
<DropdownSelect
  ariaLabel="t(toolbar.label_placeholder, {\\"context\\":\\"filter\\"})"
  buttonVariant="default"
  className=""
  direction="down"
  id="generatedid-"
  isDisabled={false}
  isDropdownButton={false}
  isInline={true}
  isToggleText={true}
  maxHeight={null}
  name={null}
  onSelect={[Function]}
  options={
    Array [
      Object {
        "component": [Function],
        "selected": true,
        "title": [Function],
        "value": "search_by_name",
      },
      Object {
        "component": [Function],
        "title": [Function],
        "value": "search_credentials_by_name",
      },
      Object {
        "component": [Function],
        "title": [Function],
        "value": "source_type",
      },
    ]
  }
  placeholder="t(toolbar.label_placeholder, {\\"context\\":\\"filter\\"})"
  position="left"
  selectedOptions={null}
  splitButtonVariant={null}
  toggleIcon={
    <FilterIcon
      color="currentColor"
      noVerticalAlign={false}
      size="sm"
    />
  }
  variant="single"
/>
`;
